File: backend/app/services/family_service.py
from sqlalchemy.orm import Session
from sqlalchemy import select
from app.models.models import Family, FamilyMember, User
from app.models.schemas import FamilyCreate, FamilyResponse
import uuid
import random
import string


class FamilyService:
    """Family management service"""

    @staticmethod
    def generate_invite_code() -> str:
        """Generate a unique invite code"""
        return "".join(random.choices(string.ascii_uppercase + string.digits, k=8))

    @staticmethod
    def create_family(db: Session, user: User, family_create: FamilyCreate) -> Family:
        """Create a new family"""
        # Generate unique invite code
        invite_code = FamilyService.generate_invite_code()
        while db.query(Family).filter(Family.invite_code == invite_code).first():
            invite_code = FamilyService.generate_invite_code()

        # Create family
        family = Family(
            id=str(uuid.uuid4()),
            name=family_create.name,
            description=family_create.description,
            invite_code=invite_code,
            tv_name=family_create.tv_name,
        )
        db.add(family)
        db.flush()

        # Add creator as admin member
        member = FamilyMember(
            id=str(uuid.uuid4()),
            user_id=user.id,
            family_id=family.id,
            role="admin",
        )
        db.add(member)
        db.commit()
        db.refresh(family)

        return family

    @staticmethod
    def get_family_members(db: Session, family_id: str) -> list:
        """Get all members of a family"""
        members = (
            db.query(FamilyMember).filter(FamilyMember.family_id == family_id).all()
        )

        # Enrich with user data
        result = []
        for member in members:
            user = db.query(User).filter(User.id == member.user_id).first()
            if user:
                result.append(
                    {
                        "id": member.id,
                        "uid": user.uid,
                        "name": user.name,
                        "email": user.email,
                        "role": member.role,
                        "avatar": member.avatar,
                        "joinedAt": member.joined_at,
                    }
                )

        return result

    @staticmethod
    def get_user_families(db: Session, user: User) -> list:
        """Get all families for a user"""
        family_members = (
            db.query(FamilyMember).filter(FamilyMember.user_id == user.id).all()
        )

        families = []
        for fm in family_members:
            family = db.query(Family).filter(Family.id == fm.family_id).first()
            if family:
                families.append(
                    {
                        "id": family.id,
                        "name": family.name,
                        "description": family.description,
                        "inviteCode": family.invite_code,
                        "tvName": family.tv_name,
                        "createdAt": family.created_at,
                    }
                )

        return families

    @staticmethod
    def join_family(db: Session, user: User, invite_code: str) -> Family:
        """Join a family with invite code"""
        family = db.query(Family).filter(Family.invite_code == invite_code).first()
        if not family:
            raise ValueError("Invalid invite code")

        # Check if user is already a member
        existing = (
            db.query(FamilyMember)
            .filter(
                FamilyMember.family_id == family.id,
                FamilyMember.user_id == user.id,
            )
            .first()
        )
        if existing:
            raise ValueError("Already a member of this family")

        # Add user to family
        member = FamilyMember(
            id=str(uuid.uuid4()),
            user_id=user.id,
            family_id=family.id,
            role="member",
        )
        db.add(member)
        db.commit()

        return family

    @staticmethod
    def invite_member(db: Session, family_id: str, email: str):
        """Invite a member by email (placeholder for email service)"""
        family = db.query(Family).filter(Family.id == family_id).first()
        if not family:
            raise ValueError("Family not found")

        # In production, would send email with invite link
        # For now, just return success
        return {"message": f"Invite sent to {email}"}

    @staticmethod
    def get_invite_code(db: Session, family_id: str) -> str:
        """Get the invite code for a family"""
        family = db.query(Family).filter(Family.id == family_id).first()
        if not family:
            raise ValueError("Family not found")

        return family.invite_code

File: frontend/src/services/familyService.js
import api from './apiService';
import { getProductFamilies, getProductFamilyMembers } from '../data/productData.js';

class FamilyService {
  constructor() {
    this.families = [];
    this.currentFamily = null;
    this.familyMembers = [];
    this.listeners = [];
  }

  onFamilyChange(callback) {
    this.listeners.push(callback);
    return () => {
      this.listeners = this.listeners.filter(l => l !== callback);
    };
  }

  notifyListeners() {
    this.listeners.forEach(callback => callback({
      families: this.families,
      currentFamily: this.currentFamily,
      members: this.familyMembers
    }));
  }

  // Get all families for current user
  async getUserFamilies() {
    try {
      const response = await api.get('/users/families');
      this.families = response.data.families || response.data;
      if (this.families.length > 0 && !this.currentFamily) {
        this.setCurrentFamily(this.families[0]);
      }
      this.notifyListeners();
      return this.families;
    } catch (error) {
      console.warn('Using local family data:', error.message);
      this.families = getProductFamilies();
      if (this.families.length > 0 && !this.currentFamily) {
        this.setCurrentFamily(this.families[0]);
      }
      this.notifyListeners();
      return this.families;
    }
  }

  // Create a new family
  async createFamily(name, description = '') {
    try {
      const response = await api.post('/families', {
        name,
        description
      });
      const newFamily = response.data;
      this.families.push(newFamily);
      this.setCurrentFamily(newFamily);
      return newFamily;
    } catch (error) {
      throw new Error(error.response?.data?.detail || 'Failed to create family');
    }
  }

  // Join a family with invite code
  async joinFamily(inviteCode) {
    try {
      const response = await api.post('/families/join', {
        invite_code: inviteCode
      });
      const family = response.data.family;
      this.families.push(family);
      this.setCurrentFamily(family);
      return family;
    } catch (error) {
      throw new Error(error.response?.data?.detail || 'Failed to join family');
    }
  }

  // Get family members
  async getFamilyMembers(familyId) {
    try {
      const response = await api.get(`/families/${familyId}/members`);
      this.familyMembers = response.data.members || response.data;
      this.notifyListeners();
      return this.familyMembers;
    } catch (error) {
      console.warn('Using local family member data:', error.message);
      this.familyMembers = getProductFamilyMembers();
      this.notifyListeners();
      return this.familyMembers;
    }
  }

  // Invite a member to family
  async inviteMember(familyId, email) {
    try {
      const response = await api.post(`/families/${familyId}/invite`, {
        email
      });
      return response.data;
    } catch (error) {
      throw new Error(error.response?.data?.detail || 'Failed to invite member');
    }
  }

  // Get invite code for family
  async getInviteCode(familyId) {
    try {
      const response = await api.get(`/families/${familyId}/invite-code`);
      return response.data.invite_code;
    } catch (error) {
      throw new Error(error.response?.data?.detail || 'Failed to get invite code');
    }
  }

  // Set current family
  setCurrentFamily(family) {
    this.currentFamily = family;
    localStorage.setItem('watchwindow_current_family', JSON.stringify(family));
    this.notifyListeners();
  }

  // Get current family
  getCurrentFamily() {
    if (!this.currentFamily) {
      const stored = localStorage.getItem('watchwindow_current_family');
      if (stored) {
        this.currentFamily = JSON.parse(stored);
      }
    }
    return this.currentFamily;
  }

  // Get current family members
  getFamilyMembers() {
    return this.familyMembers;
  }
}

export default new FamilyService();